### ISSUE TITLE
Support cmorizing MPAS data from E3SM v2 simulation

### ISSUE THREAD

**Issue by chengzhuzhang** [E3SM-Project/e3sm_to_cmip#149]
@xylar @strandwg @TonyB9000 @czender @lukaszlacinski @tomvothecoder 

Issue:

Additional mapping files and code changes needs to make to support cmorizing MPAS output (SImon Omon)

example usage for v1:
```
e3sm_to_cmip  -s  --realm mpaso  --precheck PATH_TO_existing_output_if_you_want_to_avoid_regeneration[1] \
              -v masscello  --tables-path PATH_TO_cmip6-cmor-tables/Tables/[2]  \
              --user-metadata PATH_TO_the_correct_metadata/ssp585_r1i1p1f1.json[3] \
              --map PATH_to_the_correct_map/map_oEC60to30v3_to_cmip6_180x360_aave.20181001.nc \
              --logdir PATH_TO_log_output \
              --output PATH_TO_file_output \
              --input PATH_TO_all_the_native_data_for/mpaso_segment_2015_2024[4] \
              --timeout when_to_give_up
```
Some instructions as follows:
[1] unnecessary if doing only one variable and you WANT a new output anyway.
[2] seems essentially constant - should be defaulted.
[3] you *should* have fixed metadata per case/type, so you would hope that selection can be automated by the calling script
[4] for MPAS stuff, this directory must include the required namelist, restart, and mappings files, since there is no other way to specify them
The files for v1:
mpaso.rst.1851-01-01_00000.nc
mpaso_in
oEC60to30v3_Atlantic_region_and_southern_transect.nc
map_oEC60to30v3_to_cmip6_180x360_aave.20181001.nc

**Comment by chengzhuzhang** [comment]
Based on @xylar's input: 

We will need new mapping files:
For the WC simulation we are processing, we need mapping file for EC30th60E2r2 to CMIP (following the convention of map_oEC60to30v3_to_cmip6_180x360_aave.20181001.nc)
The MOC region mask for oEC60to30v3. available [here](https://web.lcrc.anl.gov/public/e3sm/diagnostics/mpas_analysis/region_masks/EC30to60E2r2_mocBasinsAndTransects20210623.nc)

Potential code changes:
Check if the name of above input files are hard-coded in e2c and if a switch is needed to support v2 version of these files: 
- mpaso.rst.1851-01-01_00000.nc
- mpaso_in
- oEC60to30v3_Atlantic_region_and_southern_transect.nc
- map_oEC60to30v3_to_cmip6_180x360_aave.20181001.nc

Note that the case name is included in MPAS output, should check that as well.

**Comment by xylar** [comment]
Here is [map_EC30to60E2r2_to_1.0x1.0degree_conserve.nc](https://web.lcrc.anl.gov/public/e3sm/diagnostics/mpas_analysis/maps/map_EC30to60E2r2_to_1.0x1.0degree_conserve.nc) to tide you over until we have a more standardized approach for creating the mapping files you need.  Can you let me know if this is what you need?

It can be reproduce on an Anvil compute node with E3SM-Unified by running the following script.  The scrip files and ncremap call are handled through the [pyremap](https://github.com/MPAS-Dev/pyremap) tool.

``` python
#!/usr/bin/env python
import xarray
from pyremap import MpasMeshDescriptor, Remapper, get_lat_lon_descriptor

# replace with the MPAS mesh name
inGridName = 'EC30to60E2r2'

# replace with the path to the desired mesh or restart file
# As an example, use:
# https://web.lcrc.anl.gov/public/e3sm/inputdata/ocn/mpas-o/EC30to60E2r2/ocean.EC30to60E2r2.210210.nc
inGridFileName = 'ocean.EC30to60E2r2.210210.nc'

inDescriptor = MpasMeshDescriptor(inGridFileName, inGridName)

# modify the resolution of the global lat-lon grid as desired
outDescriptor = get_lat_lon_descriptor(dLon=1., dLat=1.)
outGridName = outDescriptor.meshName

mappingFileName = 'map_{}_to_{}_conserve.nc'.format(inGridName, outGridName)


remapper = Remapper(inDescriptor, outDescriptor, mappingFileName)

remapper.build_mapping_file(method='conserve', esmf_parallel_exec='srun',
                            mpiTasks=36)
```

**Comment by TonyB9000** [comment]
I thought I'd made a separate issue of this - but "--precheck" does not distinguish variant label, causing e3sm_to_cmip to refuse to process r2, r3, r4, ... if it finds "r1" data, falsely treating the data as "already generated".

Also - I assume that the namefiles and restarts provided in the native data are appropriate to generate the CMIP6 MPAS data.  But can anyone provide the scope of a namefile?  (Must it be ensemble-specific, or experiment-specific, or just model-version-specific).  Same question for "restarts" (which are not being used for restarts, but just to supply the mesh data, I've been told.)

**Comment by xylar** [comment]
> I thought I'd made a separate issue of this - but "--precheck" does not distinguish variant label, causing e3sm_to_cmip to refuse to process r2, r3, r4, ... if it finds "r1" data, falsely treating the data as "already generated".

If there isn't already one, I'd make a separate issue of that.  Definitely should be fixed or `--precheck` is not very useful.

**Comment by xylar** [comment]
> Also - I assume that the namefiles and restarts provided in the native data are appropriate to generate the CMIP6 MPAS data. But can anyone provide the scope of a namefile? (Must it be ensemble-specific, or experiment-specific, or just model-version-specific). Same question for "restarts" (which are not being used for restarts, but just to supply the mesh data, I've been told.)

The namelist file is configuration options so it doesn't have a concept of what mesh it's running on necessarily.  The restart file does, of course.

It would be best if both the namelist file and restart file are from the simulation being processed.  While it is unlikely, it is possible that the namelist options used by e2c might differ between simulations with the same horizontal mesh (e.g. `EC30to60E2r2`).  Likewise, I believe the restart file is used in e2c to get a vertical coordinate.  This could differ between simulations with the same horizontal mesh because we sometimes use different vertical coordinates.

Does that help?

**Comment by TonyB9000** [comment]
I think so.  During the zstash "short-term archiving", namefiles and restarts are included with each experiment run (to include per ensemble, per frequency, etc).  I was hoping that some commonality would allow that the native "per realm, per frequency" publications not each require a reiterated publication of the namefile (and selected restarts), as their format does not conform to the model output (naming conventions).  Sterling has created a mock "realm = misc, frequency = fixed" category to allow publication "parallel" to the given datasets, but this had also led to issues with our automation scheme.  The fact that we need these for cmorizing means that I must extract them, and place them *somewhere* they can be found, but not mix them with the "model-output" publication data.

**Comment by TonyB9000** [comment]
@xylar When you wrote, "best if both the namelist file and restart file are from the simulation being processed", what is intended by "the simulation"?  Are different ensembles "different simulations" because they begin with different initial conditions?  Or are those considered just "different runs" of the same simulation?  Thanks!

**Comment by TonyB9000** [comment]
Ok, thanks!  That does help.

As an experiment, I extracted every instance of "mpaso_in" from our v2 1pctCO2 arechive (v2.LR.1pctCO2_0101),and renamed each as "mpaso_in.<tar_path_components>" in order to compare them.  That list is

-rw-r--r--. 1 bartoletti1 publishers 36917 Jul 25  2021 mpaso_in.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36917 Jul 25  2021 mpaso_in.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36917 Sep 16 13:00 mpaso_in-run
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.L_1x10_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.L_1x10_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.L_1x10_ndays.run
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M_1x10_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M_1x10_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M_1x10_ndays.run
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M2_1x10_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M2_1x10_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M2_1x10_ndays.run
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M80_1x10_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M80_1x10_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.M80_1x10_ndays.run
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.S_1x10_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.S_1x10_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.S_1x10_ndays.run
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.XS_1x10_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.XS_1x10_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36926 Sep 16 13:00 mpaso_in-tests.XS_1x10_ndays.run
-rw-r--r--. 1 bartoletti1 publishers 36917 Sep 16 13:00 mpaso_in-tests.XS_2x5_ndays.case_scripts.Buildconf.mpasoconf
-rw-r--r--. 1 bartoletti1 publishers 36917 Sep 16 13:00 mpaso_in-tests.XS_2x5_ndays.case_scripts.CaseDocs
-rw-r--r--. 1 bartoletti1 publishers 36917 Sep 16 13:00 mpaso_in-tests.XS_2x5_ndays.run

As you can see, they come in 2 sizes (and there are 24 instances of "mpaso_in" here, just for ONE ensemble).  The difference appears to involve the values for "config_do_restart" and "config_start_time":

diff mpaso_in-run mpaso_in-tests.L_1x10_ndays.case_scripts.Buildconf.mpasoconf

> 6c6
> <  config_do_restart = .true.
> ---
> >  config_do_restart = .false.
> 8c8
> <  config_start_time = 'file'
> ---
> >  config_start_time = '0001-01-01_0'

I may do this for other experiments (those with multiple ensembles) and across experiments.   I assume the one taken from "run/" is sufficient in any case (literally, in any "case").

Doing the same for "restarts" is far more problematic, as there are a variety of items with names like

- something.rst.something

**Comment by TonyB9000** [comment]
(git formatting is easily confused by "less_thans" and "greater_thans" and triple_hyphens...)

**Comment by TonyB9000** [comment]
Datapoint:  Our v2 archives contain 171 instances of "mpaso_in".  Curiously, there are none for piClim-control, piClim-histaer, or piClim-histall. If I skip "piControl-land" and "hist-all-xGHG-xaer" and look at only the mpaso_in found in the "run/" tar directory, there are 6 files, no two the same size:

```
-rw-r--r--. 1 bartoletti1 publishers 36917 Sep 16 15:06 mpaso_in_1pctCO2_run
-rw-r--r--. 1 bartoletti1 publishers 38387 Sep 16 15:06 mpaso_in_abrupt-4xCO2_run
-rw-rw----. 1 bartoletti1 publishers 36918 Sep 16 15:07 mpaso_in_hist-aer_run
-rw-rw----. 1 bartoletti1 publishers 36916 Sep 16 15:07 mpaso_in_hist-GHG_run
-rw-rw----. 1 bartoletti1 publishers 38411 Sep 16 15:07 mpaso_in_historical_run
-rw-r--r--. 1 bartoletti1 publishers 37384 Sep 16 15:08 mpaso_in_piControl_run
```
To be safe, it appears that I must keep each of these, "installed" under our staging/resource/namefiles in such a manner that the appropriate one is selected when generating the corresponding CMIP6 datasets.

**Comment by chengzhuzhang** [comment]
Hey @TonyB9000,  piClim-control, piClim-histaer, or piClim-histall are atmospheric runs which don't have coupled ocean, so no MPAS output.

**Comment by chengzhuzhang** [comment]
@TonyB9000 May I suggest that we leave out the operational details from this script, but focusing on having one v2 simulation work with new data and code modification in e2c?

**Comment by TonyB9000** [comment]
Hi @chengzhuzhang I totally forgot the all-atmospheric runs - makes sense no MPAS info, thanks!

Xylar's instructions are quite foreign to me.  I can have fun playing with that code, and see what happens.

Given Tom's admonition not to do "pip install ." for modified e2c in the various conda environments (but rather, wait for a release of some sort, then rebuild the environment using an appropriately-updated <env>.yml file), I don't see an easy way to test changes to e2c.  I am hoping that applying the right command line (or yaml-parameters file) will enable a successful v2 cmorizing run for Omon/SImon.

I will search for the given parameters:

- mpaso.rst.1851-01-01_00000.nc
- mpaso_in
- oEC60to30v3_Atlantic_region_and_southern_transect.nc
- map_oEC60to30v3_to_cmip6_180x360_aave.20181001.nc

in the e2c code, to see about how to accommodate alternatives.  These should not be hard-coded anywhere.

**Comment by TonyB9000** [comment]
BTW, The e2c command line allows:

--map <map_mpas_to_std_grid>
                        The path to an mpas remapping file. Required if realm is mpaso or mpassi.

But I think the cmor_module simply expects certain "other parameter" files (mpaso_in, mpaso.rst*, "transect.nc" to be provided in the (temp) input directories.  The issue is having these available (somewhere) so the cwl_workflows can do their magic to stuff what is needed into the input directories.

**Comment by chengzhuzhang** [comment]
> Given Tom's admonition not to do "pip install ." for modified e2c in the various conda environments (but rather, wait for a release of some sort, then rebuild the environment using an appropriately-updated .yml file), I don't see an easy way to test changes to e2c. I am hoping that applying the right command line (or yaml-parameters file) will enable a successful v2 cmorizing run for Omon/SImon.
> 
Yes, a development environment for e2c should be used here. I think Tom's suggestion only apply to use released e2c version for operational use in `datasm`. And we the right order to make things work should be using the command line, then the cwl_worklfows and then datasm.

**Comment by TonyB9000** [comment]
OK, understood.  I tend to use the cwl_workflows as a "model" to understand the pieces, from which I generate the first cut for a manual script.  (For non-MPAS, multiple calls are involved - ncremap + e2c, etc)  Also, the datasm is sufficiently verbose (now) that missing things become visible.

Now that I've pulled down the "masks and maps", and put into place the various mpaso_in files, I should be able to reference them manually and see what happens.

**Comment by czender** [comment]
Hi @TonyB9000,

In answer to some of the earlier questions above:
@chengzhuzhang Yes, CMIP6 has grid requirements so we need to be careful the 1x1 grid is as follows: 
Latitudes are centered at cell midpoints and poles are at corners/edges of first and last gridcells (in south->north order). Longitudes are centered at cell midpoints and Greenwich is at west edge (not center!) of first longitude cell. That grid can be re-made from scratch or you can use `~zender/data/grids/cmip6_180x360_scrip.20181001.nc`:
```
zender@compy-e:~/data/grids$ ls -l cmip6_180x360_scrip.20181001.nc 
-rw-r--r-- 1 zender users 5963880 Sep 10  2018 cmip6_180x360_scrip.20181001.nc
```
The corresponding map-files that we have been using are also on all the standard machines, e.g.:
```
zender@compy-e:~/data/maps$ ls -l map_EC30to60E2r2_to_cmip6*
-rw-r--r-- 1 zender users 49713840 Feb 17  2022 map_EC30to60E2r2_to_cmip6_180x360_aave.20220301.nc
-rw-r--r-- 1 zender users 49738141 Feb 17  2022 map_EC30to60E2r2_to_cmip6_180x360_mono.20220301.nc
-rw-r--r-- 1 zender users 49725856 Feb 17  2022 map_EC30to60E2r2_to_cmip6_180x360_nco.20220301.nc
```
The instructions @xylar provided should re-produce the `aave` map assuming the SCRIP file https://web.lcrc.anl.gov/public/e3sm/inputdata/ocn/mpas-o/EC30to60E2r2/EC30to60E2r2.scrip.201005.nc has the same grid as the full data file
https://web.lcrc.anl.gov/public/e3sm/inputdata/ocn/mpas-o/EC30to60E2r2/ocean.EC30to60E2r2.210210.nc
The commands used to create these maps are:
```
ncremap --alg_typ=conserve --grd_src=${DATA}/grids/EC30to60E2r2.scrip.201005.nc --grd_dst=${DATA}/grids/cmip6_180x360_scrip.20181001.nc --map=${DATA}/maps/map_EC30to60E2r2_to_cmip6_180x360_aave.20220301.nc
ncremap --alg_typ=fv2fv_flx --grd_src=${DATA}/grids/EC30to60E2r2.scrip.201005.nc --grd_dst=${DATA}/grids/cmip6_180x360_scrip.20181001.nc --map=${DATA}/maps/map_EC30to60E2r2_to_cmip6_180x360_mono.20220301.nc
ncremap --alg_typ=nco --grd_src=${DATA}/grids/EC30to60E2r2.scrip.201005.nc --grd_dst=${DATA}/grids/cmip6_180x360_scrip.20181001.nc --map=${DATA}/maps/map_EC30to60E2r2_to_cmip6_180x360_nco.20220301.nc
```
Any map-checker (e.g., `ncks --chk_map map.nc`) will show that the TR `mono` map is superior to the ESMF and NCO maps. The ESMF and NCO maps are roughly equal to eachother in quality, with each doing slightly better or worse than the other depending on the exact statistic. For v1 E3SM used ESMF maps for both online (simulation) and offline (post-processing) work. My understanding is that in v2 E3SM switched to TR FV->FV maps (`mono`) for fluxes and kept ESMF `bilin` for state variables. It's hard to use exactly what the forward model used for all fields because the variables are all in one file yet the maps that produced them are field-dependent (flux vs. state). And we used `aave` for CMIP6 so it's probably OK to use that going forward. However, @rljacob and @mt5555 should probably weigh-in on whether E3SM should produce its "official" v2 contributions to MIPs with ESMF `aave` or TR `mono` maps (or something else) so we get started on the right foot and don't have to change that later.

**Comment by xylar** [comment]
@TonyB9000

> I may do this for other experiments (those with multiple ensembles) and across experiments. I assume the one taken from "run/" is sufficient in any case (literally, in any "case").

Yes, only the `mpaso_in` (and `mpassi_in` for sea ice) from the `run` directory are needed.  I don't know what the deal is with all the other ones you found.

> Doing the same for "restarts" is far more problematic, as there are a variety of items with names like
something.rst.something

We want any file in the `run` directory named `mpaso.rst.<date_time>.nc`.  We do not want any of the other restart files that are related to other kinds of output such as history files.

> Xylar's instructions are quite foreign to me. I can have fun playing with that code, and see what happens.

I'm happy to try to clarify if you can be more specific about what you didn't follow.

**Comment by xylar** [comment]
I fully advocate using @czender's maps and generating them with `ncremap`. I can add all 3 of @czender's `ncremap` calls to the `compass` tools so we generate these mapping files automatically with each new MPAS-Ocean and -Seaice mesh.  I can make sure they end up in standard `diagnostics/maps` directory that I believe we have agreed on.

**Comment by tomvothecoder** [comment]
> > Given Tom's admonition not to do "pip install ." for modified e2c in the various conda environments (but rather, wait for a release of some sort, then rebuild the environment using an appropriately-updated .yml file), I don't see an easy way to test changes to e2c. I am hoping that applying the right command line (or yaml-parameters file) will enable a successful v2 cmorizing run for Omon/SImon.
> 
> Yes, a development environment for e2c should be used here. I think Tom's suggestion only apply to use released e2c version for operational use in `datasm`. And we the right order to make things work should be using the command line, then the cwl_worklfows and then datasm. I forgot to tag @tomvothecoder at the first place. My bad..

Jill is correct here. Installing and testing a local build with code changes using `pip install .` is expected for the **e2c development env**. However, `pip install .` is not intended for production usage (aka `datasm` envs).

I also agree with Jill's order for code changes (using the command line, then the cwl_worklfows and then datasm).

**Comment by TonyB9000** [comment]
Understood.  I tend to use the cwl_workflows as a "short-cut RTFM" to understand the expected "command-line" inputs and order (ncremap, e2c, etc).  As I haven't been making changes to e2c (yet to be tested), I HAVE been making datasm changes, expanding into areas that were left undone during the initial development.  This enabled the generation of the v2 "fx" datasets.

**Comment by TonyB9000** [comment]
@xylar 

> only the mpaso_in (and mpassi_in for sea ice) from the run directory are needed

OK.  Just to clarify - these are still experiment-specific.  I have established a (for E3SM) a global

/p/user_pub_e3sm/staging/resource/namefiles/<ModelVersion>/<Experiment>/[mpaso_in, mpassi_in]...

where each namefile was pulled from the appropriate archive (run/).  Likely, this is overkill, but I don't want to risk applying the wrong namefile in CMIP6 generation.  Likewise, I will pull the first *.rst.* restart file from each archive, to produce a similarly accessible array of restarts for the CMIP6 tools that need them (their meshes).

(Someday, I will understand what distinguishes "map", "grid", and "mesh"...)

**Comment by chengzhuzhang** [comment]
@czender Thank you for your input even during travel time. (and I'm glad you finally got your compy account approved!)
Good to know that the new mapping files already exist. For post-processing data for CMIP, it is probably okay to use either `mono` from TR or `aave` from ESMF. I'm asking Tony to pull both files, but will see if @rljacob or @mt5555 has a preference. 

@xylar Thanks for the reminder for centralizing the map files for post-processing and making them available from the data server. That way, developers like @strandwg who don't have access to a E3SM machine  can have access.

**Comment by xylar** [comment]
@TonyB9000, what you say in https://github.com/E3SM-Project/e3sm_to_cmip/issues/149#issuecomment-1251357655 sounds right to me.

> (Someday, I will understand what distinguishes "map", "grid", and "mesh"...)

A "map" goes from one grid or mesh to another.  A grid is regular (like latitude-longitude or a box with x and y or something like that).  A "mesh" is unstructured (more irregular).  MPAS components use meshes made up of hexagons and pentagons that can't be that don't line up on a nice grid.

**Comment by TonyB9000** [comment]
@xylar Thanks!  (I can't imagine the difficulty mapping values across irregular meshes - averaging or interpolating reasonably.)  So I assume "masks" also have an implied grid or mesh, and supply (or mask) values to a corresponding grid or mesh.   Naming conventions?

Now I see that you have an "n-squared" issue generating all possible combinations of grid/mesh to grid/mesh mappings.  Still, I think it would be optimum to generate all of them and make them available, rather than trust users to generate their own as needed.  (I wrote my own codes for arbitrary rescaling and rotations of bitmap images - seems simple in comparison to dealing with irregular meshes...)

**Comment by xylar** [comment]
> Now I see that you have an "n-squared" issue generating all possible combinations of grid/mesh to grid/mesh mappings. Still, I think it would be optimum to generate all of them and make them available, rather than trust users to generate their own as needed.

We try to be reasonable about which combinations are likely to be needed and support only those.  In the case of MPAS mesh to 1 degree lat/lon, that's a pretty obvious one.  But as @czender points out, there are still many choices about how you do that.  Nearest neighbor? Bilinear interpolation? Carefully compute the polygons of overlap between each grid cell on each grid and sum them up (conservative)?  In the last case, do you keep values constant within a polygon or allow it to vary?  The TempestRemap approach he was speaking of is this most sophisticated option.  It also takes the longest to compute the mapping file.

**Comment by TonyB9000** [comment]
Yes, I imagine varied weighting schemes exist.  When downscaling I applied proportional averaging.  Upscaling, I used distance to nearest source neighbors - also for rotations (easier than calculating what fraction of a rotated rectangle overlaps another.)

I always wanted to try different schemes and judge them, by translating back and forth to see how close to "unchanged" one could get. (upscale/downscale, rotate left/right, etc)  (how I spent my summer vacation...)

**Comment by chengzhuzhang** [comment]
Hey@xylar, using the new mapping file/MOC region mask, @TonyB9000 made some nice progress to modify e2c and generated one of the Omon variable: fsitherm. But we don't know how to verify if the generated time series make sense. Would you advice what variable(s) we should provide for verification and would you help to look at the file.

Another question: is there a naming convention for the MOC region mask file? 
for v1: oEC60to30v3_Atlantic_region_and_southern_transect.nc
for v2: EC30to60E2r2_mocBasinsAndTransects20210623.nc
We need to rename the v2 file so that e2c can pick it up. 

Thanks!

**Comment by xylar** [comment]
I won't have time to help with this until the Hackathon is over (so next week).  But I should be able to take a look after that.

The MOC masks for v2 are all named `<MPAS_mesh>_mocBasinsAndTransects20210623.nc`.  The date indicates when the regions were defined so it should be the same for all meshes, regardless of when they were defined.  These masks shoudl be available in `diagnostics/mpas_analysis/region_masks` but let me know if any are missing.

**Comment by xylar** [comment]
@darincomeau and @milenaveneziani, could you advise us on what variable(s) would be needed on the sea-ice side to compute `fsitherm`?
https://github.com/PCMDI/cmip5-cmor-tables/blob/master/Tables/CMIP5_Omon#L3945-L3964
"Water Flux into Sea Water due to Sea Ice Thermodynamics"

I believe this should be the sum of melting and freezing

**Comment by xylar** [comment]
@darincomeau said on slack that he thinks `fsitherm` likely corresponds to `timeMonthly_avg_seaIceFreshWaterFlux` from the MPAS-Ocean `timeSeriesStatsMonthly` files.  @TonyB9000, is that what you used?  If so, it should be a straightforward remap from the MPAS-Ocean grid to the lat-lon grid using the conservative mapping file (since this is a flux).  I can plot the original MPAS-Ocean data and the remapped data side by side if you point me to the specific output you used and the results you got.  Again, probably not until next week.

**Comment by TonyB9000** [comment]
Thanks Xylar!  No hurry about any of this.  I have scores of variables to cmorize, and fsitherm just happened to be one that we are using for testing.  But having a solid naming standard for grids, mapfiles, and masks (and a means of knowing which apply in what circumstance) is something I struggle to grasp.

**Comment by chengzhuzhang** [comment]
Resolved by 
#151 #154 for general code change accommodating v2 file name convention 
and #155 for handling Sea-ice SGS weighting with `ncremap`
